pkg/operator/ceph/cluster/controller.go

```diff
@@ -133,6 +162,37 @@ func (c *ClusterController) StopWatch() {
 // ************************************************************************************************
 // Add event functions
 // ************************************************************************************************
+func (c *ClusterController) onK8sNodeAdd(obj interface{}) {
+	newNode, ok := obj.(*v1.Node)
+	if !ok {
+		logger.Warningf("Expected NodeList but handler received %#v", obj)
+	}
+	for i := range newNode.Spec.Taints {
+		if newNode.Spec.Taints[i].Effect == "NoSchedule" {
+			logger.Debugf("Skipping -> Node %s is Unschedulable", newNode.Labels[apis.LabelHostname])
+			return
+		}
+	}
+
+	for _, cluster := range c.clusterMap {
+		if cluster.Spec.Storage.AnyUseAllDevices() == false {
+			logger.Debugf("Skipping -> Do not use all Devices")
+			continue
+		}
+
+		if valid, _ := k8sutil.ValidNode(*newNode, cluster.Spec.Placement.All()); valid == true {
+			logger.Debugf("Adding %s to cluster %s", newNode.Labels[apis.LabelHostname], cluster.Namespace)
+			err := cluster.createInstance(c.rookImage)
```

Comment:
I'm wondering if we will need to protect against some race conditions when calling `createInstance()`
- When the operator starts, it already is calling `createInstance()`. If we get these events for new nodes every time the operator starts, can we ignore them at startup? 
-`createInstance()` could be running because the CRD was updated and it can take many minutes or even more than an hour to complete for large clusters. If a node is added while that orchestration is ongoing, we should delay calling `createInstance()` again until the first instance is completed. Maybe we need a request queue, or maybe we should just block this call.

Reply:
This is a great point. I also wonder what would happen if I added node D to a cluster, then updated node D while the operator's `createInstance` was still running. Will there be issues if we process both? Or if they're processed out of order? Can we just process the last one to keep updates to a minimum? 

@travisn what would happen if the operator restarted before the add/update could be processed? Does the operator check nodes on startup?

It seems to me that a request queue might be the safest approach for the time being, with optimizations added later.

Reply:
Each time the operator starts it runs through a full orchestration to make sure the desired state of the CRDs is applied. If the operator had been restarted before the previous orchestration was done, the operator should again attempt to complete that change simply because it is trying to fully apply desired state. 

Instead of a queue, I wonder if we could get away with a simple bool such as `orchestrationPending`. When the `cluster.createInstance()` completes, it could start over if that bool is true. I don't think we need to queue multiple requests since each call to `createInstance()` should be a full orchestration. This would assume we're not caching anything specific to the new event such as which node was added. The orchestration would just notice and pick it up already.